src/Map.ts

```diff
@@ -0,0 +1,465 @@
+ *
+ * @category Transformations
+ *
+ * @returns A new `Map` with the added key/value pair
+ */
+const add =
+    <K, V>(
+        [key, value]: readonly [K, V],
+        equalityComparer: EqualityComparer<K> = defaultEqualityComparer
+    ) =>
+    (map: ReadonlyMap<K, V>): ReadonlyMap<K, V> => {
+        if (map.size < 1) {
+            const out = empty<K, V>()
+            out.set(key, value)
+            return out
+        }
+
+        const copy = new globalThis.Map(map)
+
+        return pipe(
+            map,
+            findWithKey(key, equalityComparer),
+            Option.match({
+                none: () => {
+                    copy.set(key, value)
+                    return copy
+                },
+                some: ([k]) => {
+                    copy.set(k, value)
+                    return copy
+                },
+            })
+        )
+    }
+
+/**
+ * Make a new `Map` by producing a new each value for each key using
+ * the given function.
+ *
+ * @category Mapping
+ * @category Transformations
+ */
+const map =
+    <K, V, R>(f: (k: K, v: V) => R) =>
+    (map: ReadonlyMap<K, V>): ReadonlyMap<K, R> => {
+        if (map.size < 1) {
+            return empty<K, R>()
+        }
+
+        const out = empty<K, R>()
+
+        for (const [k, v] of map) {
+            out.set(k, f(k, v))
+        }
+
+        return out
+    }
+
+/**
+ * Get the first key for which the given predicate function returns
+ * true, wrapped in a `Some`. If no key is found, returns `None`. Uses
+ * the given `OrderingComparer` if passed, otherwise defaults to default
+ * ASCII-based sort.
+ *
+ * @category Lookups
+ */
+const findKey =
+    <K extends {}>(
+        predicate: Predicate<K>,
+        orderingComparer: OrderingComparer<K> = defaultOrderingComparer
+    ) =>
+    <V>(map: ReadonlyMap<K, V>): Option<K> =>
+        Option.ofNullish(keys(orderingComparer)(map).find(predicate))
+
+/**
+ * Creates a new empty map. Essentially an alias for `new globalThis.Map()`.
+ * Provided for convience to avoid having to use `globalThis`.
+ *
+ * @category Constructors
+ */
+const empty = <K = never, V = never>() => new globalThis.Map<K, V>()
+
+/**
+ * Returns `true` if at least one _value_ in the `Map` returns `true`
+ * for the given predicate function.
+ *
+ * @category Lookups
+ * @category Utils
+ */
+const exists =
+    <V>(predicate: Predicate<V>) =>
+    <K>(map: ReadonlyMap<K, V>): boolean => {
+        if (map.size < 1) {
+            return false
+        }
+
+        for (const [, v] of map) {
+            if (predicate(v)) {
+                return true
+            }
+        }
+
+        return false
+    }
+
+/**
+ * Replace the value at a given key in the map using the given
+ * replacement function. Will use the given `EqualityComparer`
+ * if passed. Otherwise defaults to reference equality (triple equals).
+ *
+ * If the key isn't in the map, returns the map unchanged.
```

Comment:
I might argue we have two implementations of this with this in mind. `change` and `changeOrAdd` so that we can fail fast if we were expecting the key to exist. aside: thoughts on `update` instead of `change`?

Reply:
`update` could make sense. although, I think `change` still has the edge in my mind because `update` would imply to me that I'm just giving it a new value at that key. `change` I think makes more sense here because we're altering the value at the key using a "change" function.

And, I think `add` currently works like you suggested (`changeOrAdd`). It will add a new key or update the existing one. I believe fp-ts calls this `upsertAt`

Reply:
>And, I think add currently works like you suggested (changeOrAdd). It will add a new key or update the existing one. 

Should it? 😅 I think I'd make the same argument about allowing consumers to fail fast when one of these methods does something they don't expect. i.e. I'd expect `add` to fail if the key already exists.

Reply:
I think in this case the precedent is actually in favor of treating `add` as an `upsert`. The built-in `.set` method on maps updates the key if it already exists. And F# maps use the same behavior. I'm willing to change the name, though, to make it more clear if that's an issue.

Reply:
Will rename to `set`! Thanks for the discussion 👍

Reply:
https://github.com/taylorn16/fp-toolkit/pull/20/commits/e8100c4f0a6e046d3f6c1ab2747c58eafb776bc4